src/main/scala/midas/widgets/DMATransport.scala

```diff
@@ -0,0 +1,160 @@
+// See LICENSE for license details.
+
+package midas
+package widgets
+
+import core.{DMANastiKey}
+import junctions._
+
+import chisel3._
+import chisel3.util._
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.util.{DecoupledHelper}
+
+trait HasDMA {
+  self: Widget =>
+  val dma = IO(Flipped(new NastiIO()( p.alterPartial({ case NastiKey => p(DMANastiKey) }))))
+  // Specify the size that you want the address region to be in the DMA memory map
+  // For proper functioning, the region should be at least as big as the
+  // largest read/write request you plan to send over PCIS
+  val dmaBytes = dma.nastiXDataBits / 8
+  val dmaSize: BigInt
+}
+
+trait DMAToHostCPU extends HasDMA {
+  self: Widget =>
+  val toHostCPUQueueDepth: Int
+
+  require(dma.nastiXDataBits == 512, "DMA width must be 512 bits (PCIS)")
+
+  //val incomingPCISdat = Module(new SplitSeqQueue)
+  val outgoingPCISdat = Module(new BRAMQueue(toHostCPUQueueDepth)(UInt(dma.nastiXDataBits.W)))
+
+  // incoming/outgoing queue counts to replace ready/valid for batching
+  val outgoingCount = RegInit(0.U(32.W))
+
+  when (outgoingPCISdat.io.enq.fire() && outgoingPCISdat.io.deq.fire()) {
+    outgoingCount := outgoingCount
+  } .elsewhen (outgoingPCISdat.io.enq.fire()) {
+    outgoingCount := outgoingCount + 1.U
+  } .elsewhen (outgoingPCISdat.io.deq.fire()) {
+    outgoingCount := outgoingCount - 1.U
+  } .otherwise {
+    outgoingCount := outgoingCount
+  }
+
+  // check to see if pcis has valid output instead of waiting for timeouts
+  attach(outgoingPCISdat.io.deq.valid, "pcis_out_valid", ReadOnly)
+  attach(outgoingCount, "outgoing_count", ReadOnly)
+
+  // TODO, will these queues bottleneck us? - Biancolin: no. -> just read out of BRAM queue?
+  val ar_queue = Queue(dma.ar, 10)
```

Comment:
@sagark What's the rationale for these? They are pretty enormous, and i think additional DMA buffering (for timing closure) can be moved into FPGATop (they should also be PipeQueues i think).

Reply:
Yeah, those are leftover from the initial NIC widget implementation that didn't use DMA. They can be removed.

Reply:
I'll remove them as part of PrintSynthesis bring up.